src/What4/Serialize/FastSExpr.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE OverloadedStrings #-}
+-- | This module implements a specialized s-expression parser
+--
+-- The parser in s-cargot is very general, but that also makes it a bit
+-- inefficient.  This module implements a drop-in replacement parser for the one
+-- in What4.Serialize.Parser using megaparsec.  It is completely specialized to
+-- the types in this library.
+module What4.Serialize.FastSExpr (
+  parseSExpr
+  ) where
+
+import           Control.Applicative
+import qualified Control.Monad.Fail as MF
+import           Data.Parameterized.Some ( Some(..) )
+import           Data.Ratio ( (%) )
+import qualified Data.SCargot.Repr.WellFormed as SC
+import qualified Data.Text as T
+import qualified Data.Void as DV
+import           Numeric.Natural ( Natural )
+import qualified Text.Megaparsec as TM
+import qualified Text.Megaparsec.Char as TMC
+import qualified Text.Megaparsec.Char.Lexer as TMCL
+import qualified What4.BaseTypes as WT
+import qualified What4.Serialize.SETokens as WST
+
+-- | Parse 'T.Text' into the well-formed s-expression type from s-cargot.
+parseSExpr :: T.Text -> Either String (SC.WellFormedSExpr WST.Atom)
+parseSExpr t =
+  case TM.runParser (ws >> parse) "<input>" t of
+    Left errBundle -> Left (show errBundle)
+    Right a -> Right a
+
+type Parser a = TM.Parsec DV.Void T.Text a
+
+parse :: Parser (SC.WellFormedSExpr WST.Atom)
+parse = parseList <|> (SC.WFSAtom <$> lexeme parseAtom)
+
+parseList :: Parser (SC.WellFormedSExpr WST.Atom)
+parseList = do
+  _ <- lexeme (TMC.char '(')
+  items <- TM.many parse
+  _ <- lexeme (TMC.char ')')
+  return (SC.WFSList items)
+
+parseId :: Parser T.Text
+parseId = T.pack <$> ((:) <$> first <*> TM.many rest)
+  where
+    w4symbol c = c == '@'
+               || c == '+'
+               || c == '-'
+               || c == '='
+               || c == '<'
+               || c == '>'
+               || c == '_'
+               || c == '.'
+    first = TMC.letterChar <|> TM.satisfy w4symbol
+    rest = TMC.alphaNumChar <|> TM.satisfy w4symbol
+
+parseNat :: Parser Natural
+parseNat = do
+  _ <- TMC.string "#u"
+  TMCL.decimal
+
+parseInt :: Parser Integer
+parseInt = TMCL.decimal <|> (negate <$> (TMC.char '-' *> TMCL.decimal))
+
+parseReal :: Parser Rational
+parseReal = do
+  _ <- TMC.string "#r"
+  n <- TMCL.decimal
+  _ <- TMC.char '/'
+  d <- TMCL.decimal
+  return (n % d)
+
+parseBV :: Parser (Int, Integer)
+parseBV = do
+  _ <- TMC.char '#'
+  t <- TM.anySingle
+  case t of
+    'b' -> parseBin 0 0
+    'x' -> parseHex
+    _ -> MF.fail ("Invalid bitvector class: " ++ show t)
```

Comment:
Is the failure case necessary here?  A normal parse failure seems adequate, and I think that the `TM.try` wrapped around  this call will discard the failure message.  If the failure case was removed, this could be rewritten as something like

```
   parseBV = TMC.char '#' *> (TMC.char 'b' *> parseBin 0 0
                              <|>
                              TMC.char 'x' *> parseHex)
```

Reply:
This would also allow removal of the MF import.

Reply:
My thought there was to avoid an admittedly small amount of backtracking.  We can decide exactly which path we are on by parsing one character, rather than parsing a character, failing, and then re-parsing the character and checking if it is an 'x'

Reply:
Ahh, makes sense.